File: input_tools/version/create-tz_version.py
#!/usr/bin/python3 -B

# Copyright 2017 The Android Open Source Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Generates a time zone version file"""

import argparse
import os
import shutil
import subprocess
import sys

sys.path.append('%s/external/icu/tools' % os.environ.get('ANDROID_BUILD_TOP'))
import i18nutil

sys.path.append('%s/system/timezone' % os.environ.get('ANDROID_BUILD_TOP'))
import tzdatautil

android_build_top = i18nutil.GetAndroidRootOrDie()
android_host_out_dir = i18nutil.GetAndroidHostOutOrDie()
timezone_dir = os.path.realpath('%s/system/timezone' % android_build_top)
i18nutil.CheckDirExists(timezone_dir, 'system/timezone')

def RunCreateTzVersion(properties_file):
  # Build the libraries needed.
  tzdatautil.InvokeSoong(android_build_top, ['create_tz_version'])

  # Run the CreateTzVersion tool
  command = '%s/bin/create_tz_version' % android_host_out_dir
  subprocess.check_call([command, properties_file])


def CreateTzVersion(
    iana_version, revision, output_version_file):
  original_cwd = os.getcwd()

  i18nutil.SwitchToNewTemporaryDirectory()
  working_dir = os.getcwd()

  # Generate the properties file.
  properties_file = '%s/tzversion.properties' % working_dir
  with open(properties_file, "w") as properties:
    properties.write('rules.version=%s\n' % iana_version)
    properties.write('revision=%s\n' % revision)
    properties.write('output.version.file=%s\n' % output_version_file)

  RunCreateTzVersion(properties_file)

  os.chdir(original_cwd)


def main():
  parser = argparse.ArgumentParser()
  parser.add_argument('-iana_version', required=True,
      help='The IANA time zone rules release version, e.g. 2017b')
  parser.add_argument('-revision', type=int, required = True,
      help='Revision of the current IANA version')
  parser.add_argument('-output_version_file', required=True,
      help='The output path for the version file')
  args = parser.parse_args()

  iana_version = args.iana_version
  revision = args.revision
  output_version_file = os.path.abspath(args.output_version_file)

  CreateTzVersion(
      iana_version=iana_version,
      revision=revision,
      output_version_file=output_version_file)

  print('Version file created as %s' % output_version_file)
  sys.exit(0)


if __name__ == '__main__':
  main()

File: input_tools/version/src/main/com/android/timezone/version/tools/CreateTzVersion.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.timezone.version.tools;

import com.android.i18n.timezone.TzDataSetVersion;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.Reader;
import java.util.Properties;

/**
 * A command-line tool for creating a time zone data set version file.
 *
 * <p>Args:
 * <dl>
 *     <dt>input properties file</dt>
 * </dl>
 *
 * <p>The input properties file must have the entries:
 * <dl>
 *     <dt>rules.version</dt>
 *     <dd>The IANA rules version.</dd>
 *     <dt>revision</dt>
 *     <dd>IANA data revision (typically 1).</dd>
 *     <dt>output.version.file</dt>
 *     <dd>The location to write the version file to.</dd>
 * </dl>
 *
 * <p>The output consists of:
 * <ul>
 *     <li>A version file.</li>
 * </ul>
 */
public class CreateTzVersion {

    private CreateTzVersion() {}

    public static void main(String[] args) throws Exception {
        if (args.length != 1) {
            printUsage();
            System.exit(1);
        }
        File f = new File(args[0]);
        if (!f.exists()) {
            System.err.println("Properties file " + f + " not found");
            printUsage();
            System.exit(2);
        }
        Properties properties = loadProperties(f);
        String ianaRulesVersion = getMandatoryProperty(properties, "rules.version");
        int revision = Integer.parseInt(getMandatoryProperty(properties, "revision"));

        // Create an object to hold version metadata for the tz data.
        TzDataSetVersion tzDataSetVersion = new TzDataSetVersion(
                TzDataSetVersion.currentFormatMajorVersion(),
                TzDataSetVersion.currentFormatMinorVersion(),
                ianaRulesVersion,
                revision);
        byte[] tzDataSetVersionBytes = tzDataSetVersion.toBytes();

        File outputVersionFile = new File(getMandatoryProperty(properties, "output.version.file"));

        // Write the tz data set version file.
        try (OutputStream os = new FileOutputStream(outputVersionFile)) {
            os.write(tzDataSetVersionBytes);
        }
        System.out.println("Wrote " + outputVersionFile);
    }

    private static String getMandatoryProperty(Properties p, String propertyName) {
        String value = p.getProperty(propertyName);
        if (value == null) {
            System.out.println("Missing property: " + propertyName);
            printUsage();
            System.exit(3);
        }
        return value;
    }

    private static Properties loadProperties(File f) throws IOException {
        Properties p = new Properties();
        try (Reader reader = new InputStreamReader(new FileInputStream(f))) {
            p.load(reader);
        }
        return p;
    }

    private static void printUsage() {
        System.out.println("Usage:");
        System.out.println("\t" + CreateTzVersion.class.getName() +
                " <tzupdate.properties file>");
    }
}
